assembl/lib/clean_input.py

```diff
@@ -60,19 +59,73 @@ def _clean_html(html_value, cleaner):
             yield f
 
 
-def sanitize_html_keep(html_value, valid_tags, valid_attributes=VALID_ATTRIBUTES):
-    """Remove unwanted tags. Unlike other cleaners like bleach, keep those tag's contents."""
-    soup = BeautifulSoup(html_value)
-    for tag in soup.find_all(True):
-        if tag.name not in valid_tags:
-            tag.hidden = True
-            if tag.name in ('p', 'br'):
-                tag.append('\n')
-        else: # it might have bad attributes
-            for attr in tag.attrs.keys():
-                if attr not in valid_attributes:
-                    del tag[attr]
-    return soup.decode_contents()
+def lxml_remove_tag(tree, index):
+    # remove a tag from a tree fragment without removing its content
+    children = tree.getchildren()
+    element = children[index]
+    tree.remove(element)
+    if element.text:
+        if index:
+            text_target = children[index - 1]
+            text_target.tail = (text_target.tail or '') + element.text
+        else:
+            tree.text = (tree.text or '') + element.text
+    for sub in element.iterchildren():
+        tree.insert(index, sub)
+        index += 1
+    tail = element.tail or ''
+    if element.tag in ('p', 'br'):
+        tail += '\n'
+    if tail:
+        if index:
+            text_target = tree.getchildren()[index - 1]
+            text_target.tail = (text_target.tail or '') + tail
+        else:
+            tree.text = (tree.text or '') + tail
+    return index
+
+
+def clean_attributes(node, valid_attributes):
+    for attname in node.attrib.keys():
+        if attname not in valid_attributes:
+            del node.attrib[attname]
+
+
+def sanitize_html_rec(fragment, valid_tags, valid_attributes):
+    i = 0
+    while i < len(fragment):
+        child = fragment.getchildren()[i]
+        sanitize_html_rec(child, valid_tags, valid_attributes)
```

Comment:
I don't understand how this recursive function eventually hits the leaf condition.

Reply:
What is returned once the child's length is 0?

Reply:
It hits the leaf condition when it's on a HTML leaf and there are no children.
Nothing is returned, this does in-place surgery on the lxml nodes.